### PR TITLE
Fragmenting sent handshake messages

### PR DESCRIPTION
**Pull request by RyanGordon**
#### Description
Previously handshake messages being sent out would not be fragmented. Most handshake messages are pretty small but a couple of exceptions, such as Certificate, mean the packet size can be greater than the MTU, which then causes some packet loss and the handshake would never be able to complete. This changes the `internalSend` function to fragment handshakes to the set MTU.

We set `MTU` as a config option, which defaults to `1200`

### CODE REVIEW

**Comment by RyanGordon** [comment]
This is ready for review now. A couple of notes:

I don't _love_ the `Marshal()` interface, but it didn't really make sense to fit it anywhere else. I can't fragment before `packet.Marshal()`, because the handshake is just an in-memory struct still; we have no idea the full size of the handshake message itself. And I can't do it after because we need to manipulate the record and handshake headers to support fragmentation properly to spec. It might take a larger refactor to have a cleaner interface. Open to thoughts here.

I also realizing I needed to tackle the
`// sequenceNumber and messageSequence line up, may need to be re-evaluated`
tech debt because the possibility of fragmentation means these will not align anymore, so there is slightly more code tweaked to fix this case.

One last change was a small improvement to the way the "handshake completed signaling" works so that we don't end up in a situation where we could cause a panic calling an already-closed channel. I can split this out if desired / I may move this change to https://github.com/pion/dtls/pull/113

**Comment by Sean-Der** [comment]
I am going to spend some time with this, but I am not a big fan of how the MTU is split out all over the code base.

I need to think over it, but it would be really nice if we didn't have to change `Marshal()` but just had a single function that took care of fragmenting for us.

 I don't think there is much danger to double parse/allocate since this isn't a hot path. I do think this could start to introduce substantial code complexity. Maybe this is the only way we can do it, but I will think it over and come back!

**Comment by Sean-Der** [comment]
@RyanGordon re handshake complete stuff, merge away! Thank you so much for putting it in a smaller PR, it makes it so much easier to just worry about a single thing.

**Comment by Sean-Der** [comment]
What do you think of keeping something in memory that maps `original sequence number` -> `fragmented`

Then only when it comes time to encrypt and send we can say `We are sending N`, but `N-1`  was actually these three so we need to account for that.

This adds more state/complexity to that function. But I just think in the long run it will pay off to have something that is very contained/testable.

**Comment by RyanGordon** [comment]
I agree it would be nicer if we could keep `Marshal()` the same. I am going to think more on this but the basic complexity is the way the inheritance works with marshaling:

recordHeader.Marshal -> handshakeHeader.Marshal -> handshake.Marshal

If we reversed it, then it becomes much easier to keep the fragmenting and other stateful logic out of Marshal

i.e. very rough-cut:
```
internalSend(message) {
  var packets [][]byte
  if message.type == handshake {
    handshakes := fragment_handshake(message.handshake)
    for handshake in handshakes {
       record = ...
       packets = append(packets, record.Marshal())
   }
  } else {
   packets = [message.Marshal()]
  }

  // send packets
}

fragment_handshake(handshake) {
  content := handshake.content.Marshal()
  origHeader := handshake.header
  fragments = split(content, MTU_SIZE)
  
  handshakes = []
  for fragment in fragments {
     fragmentHeader = copy(origHeader)
     fragmentHeader.offset = ...
     fragmentHeader.len = ...
     fragmentHeader.content = fragment
     handshakes = append(handshakes, fragmentHeader)
  }
  
  return handshakes
}

```

**Comment by Sean-Der** [comment]
This is fantastic @RyanGordon!

I am really impressed how fast you rewrote/fixed up this PR. Reading through the code I like it a lot more, it is just nice that it is only in one place.

I think that is the hardest part about OpenSSL. A lot of the TLS code duplicates logic/ a lot of things are interwoven so it is hard to follow.

This is a merge from me!

**Comment by Sean-Der** [comment]
So checking out the PR I think that if we fire messages for a flight multiple times this will cause issues.

It looks like we increment the sequenceNumber every time we send a fragment. I think this a good design decision.

However if we re-run a flight multiple times the SequenceNumber will not account for that. I think the only change we need to make is to record `sequenceNumber at start of flight` and then we should be good! Then every time we re-run that flight we need to reset back.

Happy to take care of it also!

**Comment by Sean-Der** [comment]
Another thing I realized, we currently might send different packets with the same sequenceNumber.

If this flight handler has run multiple times https://github.com/pion/dtls/blob/master/server_handlers.go#L481 we don't increment the global sequenceNumber counter until we have moved onto the next flight. 

If we hit an error during handshaking https://github.com/pion/dtls/blob/master/conn.go#L391 we will reuse a sequence number that has already been used. The global sequenceNumber will just be lower than what has actually gone out.

We need to handle sequenceNumbers and replaying flights better in general. The code is confusing error prone. Maybe we should 'replay flights' by pulling them from the local handshake cache? Then we can increment the global sequenceNumber cache w/e we want.

**Comment by RyanGordon** [comment]
> So checking out the PR I think that if we fire messages for a flight multiple times this will cause issues.
> 
> It looks like we increment the sequenceNumber every time we send a fragment. That works great for the happy case.
> 
> However if we re-run a flight multiple times the SequenceNumber will not account for that. I think the only change we need to make is to record `sequenceNumber at start of flight` and then we should be good. Then every time we re-run that flight we need to reset back.
> 
> Happy to take care of it also!

Hmm, I think this is actually how it works because `c.handshakeMessageSequence` is used for the handshake header `messageSequence` field and `c.state.localSequenceNumber` is only used for the record header `sequenceNumber` field

`c.handshakeMessageSequence` is assigned based on the old logic that would correctly set the handshake header `messageSequence` to the right value for the state of the sequence, including replays.

And now `c.state.localSequenceNumber` is only used for the record and is incremented every time a record is sent so it is always unique for each sent record (even if the record contains retransmission content). At least according to RFC6347 4.1.2.6 seems to say that the record layer sequence number should always be unique or duplicate packets could be dropped by the receiver depending on the library implementation.

**Comment by Sean-Der** [comment]
@RyanGordon 

(These are just assumptions, please correct me where I am wrong!)

* https://github.com/pion/dtls/blob/bc94af06cd09f3ec7fb246e5e0a264382da4e19e/conn.go#L379 always increments the SequenceNumber. We will never have a message with the same sequence number twice.

* This means that each call to internal send will use new sequenceNumbers https://github.com/pion/dtls/blob/bc94af06cd09f3ec7fb246e5e0a264382da4e19e/client_handlers.go#L292

`pion/dtls` itself is unable to handle gaps in sequence numbers. It is very explicit about the SequenceNumber that is used for certain messages https://github.com/pion/dtls/blob/bc94af06cd09f3ec7fb246e5e0a264382da4e19e/server_handlers.go#L188

**Comment by RyanGordon** [comment]
Hmm okay, that is interesting. I need to do more research into this to make sure I am understanding the spec correctly and that this doesn't introduce breaking behavior. (Currently) My understanding is that the record "sequence number" (I believe) should always be incrementing even on re-transmit. The handshake "message sequence" is the one that should have the specific ordering and be re-used depending on the state of the handshake and used to know where in the handshake the sender is.

I will be double checking what the spec behavior _should be_, and what impact these changes  would have and what would need to be changed to avoid any impact

**Comment by Sean-Der** [comment]
Fantastic, sorry for this review being a doozy! I really appreciate how much you are doing to land this.

I would see what OpenSSL does on retransmissions (if it increments the SequenceNumber), and honestly if E2E Lossy passes I think we are in great shape. 

I really need to make a nice way for us to do E2E Lossy against OpenSSL.

**Comment by daenney** [comment]
> I really need to make a nice way for us to do E2E Lossy against OpenSSL.
I'm working on that. Once I've recovered from a 9hr timezone shift I'll submit

**Comment by Sean-Der** [comment]
> > I really need to make a nice way for us to do E2E Lossy against OpenSSL.
> > I'm working on that. Once I've recovered from a 9hr timezone shift I'll submit

Well that is just a fantastic surprise. Thanks for grabbing that!

**Comment by RyanGordon** [comment]
Ok, this is ready for a final review @Sean-Der, et al. Tests should pass but golangci-lint seems to be doing something weird (related to Go 1.13 release?)

**Comment by Sean-Der** [comment]
Nice @RyanGordon !

Mind rebasing this off of you 1.12 pin I will make sure I review this tonight! If it passes everything that is for sure going to be a merge from me.

really nice work

File: e2e/e2e_test.go [hunk 1]
```diff
@@ -232,3 +232,33 @@ func TestPionE2ESimplePSK(t *testing.T) {
 		assertE2ECommunication(cfg, cfg, serverPort, t)
 	}
 }
+
+func TestPionE2EMTUs(t *testing.T) {
```
Comment (Sean-Der) [inline e2e/e2e_test.go]: Can you introduce loss to this test as well? I want to see how it handles gaps in sequence numbers as well.

Comment (RyanGordon) [reply e2e/e2e_test.go]: Sure thing!

Comment (RyanGordon) [reply e2e/e2e_test.go]: Done!

File: client_handlers.go [hunk 2]
```diff
@@ -171,7 +172,7 @@ func clientHandshakeHandler(c *Conn) (*alert, error) {
 			if alertPtr, err := handleSingleHandshake(expectedMessages[0].data); err != nil {
 				return alertPtr, err
 			}
-			c.state.localSequenceNumber++
+			c.handshakeMessageSequence++
```
Comment (Sean-Der) [inline client_handlers.go]: I think we can easily get this out of the global state, could help readability a bit I think.

we can put `handshakeMessageSequence` on the stack, and then look at the local cache for the last flight. Or just look at what features are enabled

Comment (Sean-Der) [reply client_handlers.go]: Don't worry about doing this in this PR, merge away!
